packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts

```diff
@@ -104,22 +123,103 @@ export abstract class DeployRetrieveExecutor<
   protected abstract postOperation(
     result: DeployRetrieveResult | undefined
   ): Promise<void>;
+  protected abstract handleSourceConflictError(e: any): void;
 }
 
 export abstract class DeployExecutor<T> extends DeployRetrieveExecutor<T> {
   protected async doOperation(
     components: ComponentSet,
     token: vscode.CancellationToken
   ): Promise<DeployResult | undefined> {
+    const projectPath = getRootWorkspacePath();
+    const connection = await WorkspaceContext.getInstance().getConnection();
+
+    const sourceTracking = await SourceTrackingService.createSourceTracking(
+      projectPath,
+      connection
+    );
+    await sourceTracking.ensureLocalTracking();
+
     const operation = await components.deploy({
-      usernameOrConnection: await WorkspaceContext.getInstance().getConnection()
+      usernameOrConnection: connection
     });
 
     this.setupCancellation(operation, token);
 
     return operation.pollStatus();
   }
 
+  protected async handleSourceConflictError(e: any) {
```

Comment:
Is this like an "error error" where something has gone wrong with our handling of source conflicts? Would it be worth logging with telemetry if so?

Reply:
No, the SourceConflictError is what STL throws when it detects conflicts.  Up above in DeployRetrieveExecutor run(), when a SourceConflictError is caught, the subclasses handleSourceConflictError() method is called so it can handle the error.  In a deploy operation, the conflict view is setup and shown, and in the RetrieveExecutor it is not handled currently.

Reply:
Yeah, part of the reason is that `SourceConflictError` in STL extends `SfError` from core:
<img width="400" alt="2023-03-22_22-04-58" src="https://user-images.githubusercontent.com/46458081/227109520-afeae153-78fb-418f-a905-c9fa50120344.png">

which is [an interesting type](https://github.com/forcedotcom/sfdx-core/blob/4dbd4907f34e5c8495a21c5be31b2cf80f52a90d/src/sfError.ts#L27) in itself!:
(it can have actions that help the users get past errors)
<img width="400" alt="2023-03-22_22-11-54" src="https://user-images.githubusercontent.com/46458081/227110231-0a51f83e-f5d2-4d03-b638-430f83e3d0b1.png">